### PR TITLE
[CI] Optimize CI resource usage

### LINKED ISSUE #11670: [CI] Optimize our CircleCI machine sizes
In our CI config we only use `medium` and `xlarge` machines. There's probably does not give us the best cost/performance ratio. We should try optimizing it a bit.

See [list of all available machine types](https://circleci.com/pricing/#compute-options-table).

On the call with CircleCI people we've been recommended to try the larger Windows and macOS machines to speed up CI runs. Currently we're using `medium` instances and there are larger ones available. Not sure if that's because larger ones weren't originally available or if we just decided smaller ones are most cost-effective. We should at least try them out and compare performance and cost (if we haven't already)

Also, we have some jobs that do not require much processing and only make use of one core. For example the ones I'm adding in #12181 and #12165 but I'm sure there are more. These can very likely be switched to `small` without affecting the total runtime too much.

Finally, we should make sure that all the linux jobs that use `xlarge` actually do benefit from it.

### PR DESCRIPTION
**Pull request by cameel**
Fixes #11670.
~Depends on #12214 (draft until that PR is merged).~ Merged.

This PR applies my conclusions from https://github.com/ethereum/solidity/issues/11670#issuecomment-954297346.

Also removes compilation-only runs for some external tests (we have them only because they're supposed to be much faster than a full test run but in some cases they actually aren't).

### CODE REVIEW

**Comment by cameel** [comment]
Here's a comparison of execution times before and after this PR:

| job | [`develop`](https://app.circleci.com/pipelines/github/ethereum/solidity/20142/workflows/cf274899-9121-4563-b422-c826d26cbdb1) | [base branch](https://app.circleci.com/pipelines/github/ethereum/solidity/20146/workflows/dcaf7643-a91a-474c-8b03-10e81620d7a9) | [this branch](https://app.circleci.com/pipelines/github/ethereum/solidity/20145/workflows/928613f9-3732-415e-b890-78573ce66087) (1) | this branch (2) | this branch (3) | this branch (4) |
|-----|-----|-----|-----|-----|-----|-----|
| `b_archlinux`                     | 18m 04s  | 19m 41s | 13m 44s | 12m 47s |  14m 17s | 12m 40s |
| `b_bytecode_ems`                  |  5m 15s  |  5m 24s |  5m 27s |  6m 11s |   5m 53s |  5m 36s |
| `b_bytecode_osx`                  | 13m 31s  | 13m 30s | 12m 53s | 14m 31s |  13m 20s | 13m 36s |
| `b_bytecode_ubu`                  |  8m 47s  |  8m 26s |  8m 38s |  8m 27s |   8m 33s |  8m 17s |
| `b_bytecode_win`                  | 10m 09s  |  9m 46s | 10m 05s | 10m 49s |  10m 15s | 10m 03s |
| `b_docs`                          |     43s  |     46s |     45s |     57s |      48s |     52s |
| `b_ems`                           |  6m 30s  |  6m 42s |  9m 19s | 10m 44s |   7m 51s |  6m 40s |
| `b_osx`                           | 19m 53s  | 20m 31s | 18m 39s | 22m 20s |  18m 38s | 19m 10s |
| `b_ubu_clang`                     |  5m 17s  |  5m 10s |  9m 16s |  9m 03s |   7m 28s |  7m 31s |
| `b_ubu_cxx20`                     |  6m 38s  |  5m 50s |  9m 46s | 11m 28s |   7m 15s |  7m 51s |
| `b_ubu_ossfuzz`                   | 16m 24s  | 15m 27s | 15m 24s | 15m 42s |  18m 43s | 17m 01s |
| `b_ubu_release`                   |  6m 46s  |  6m 30s |  6m 31s |  8m 08s |   7m 17s |  6m 31s |
| `b_ubu_static`                    |  6m 19s  |  7m 52s | 15m 34s | 16m 56s |    CRASH | 13m 16s |
| `b_ubu`                           |  7m 03s  |  6m 33s |  6m 13s |  6m 04s |   6m 35s |  5m 56s |
| `b_win_release`                   | 26m 14s  | 28m 17s | 16m 31s | 18m 57s |  17m 19s | 16m 24s |
| `b_win`                           | 29m 09s  | 27m 54s | 17m 03s | 16m 26s |  16m 31s | 16m 59s |
| `chk_antlr_grammar`               |  7m 40s  |  7m 50s |  8m 24s |  9m 13s |   8m 44s |  9m 12s |
| `chk_buglist`                     |     14s  |     14s |     15s |     16s |      24s |     17s |
| `chk_coding_style`                |     26s  |     30s |     22s |     23s |      28s |     26s |
| `chk_docs_pragma_min_version`     |     34s  |     26s |     44s |     32s |      31s |     20s |
| `chk_errorcodes`                  |     9s   |      7s |     12s |     13s |      14s |     10s |
| `chk_proofs`                      |     17s  |     14s |     23s |     18s |      18s |     22s |
| `chk_pylint`                      |     36s  |     50s |     28s |     37s |      36s |     36s |
| `chk_spelling`                    |     11s  |     14s |     23s |     26s |      20s |     16s |
| `t_archlinux_soltest`             | 13m 19s  | 14m 21s | 12m 42s | 14m 03s |  13m 56s | 14m 45s |
| `t_bytecode_compare`              |     44s  |     55s |     33s |     49s |      44s |     33s |
| `t_ems_compile_ext_colony`        |  2m 32s  |  2m 15s |  2m 40s |  3m 05s |   2m 49s |  2m 37s |
| `t_ems_compile_ext_ens`           |  2m 08s  |  2m 03s |       - |       - |        - |       - |
| `t_ems_compile_ext_gnosis_v2`     |  3m 48s  |  3m 43s |       - |       - |        - |       - |
| `t_ems_compile_ext_gnosis`        |  3m 47s  |  3m 21s |  3m 27s |  3m 39s |   4m 25s |  3m 31s |
| `t_ems_compile_ext_zeppelin`      | 11m 53s  |  5m 28s |       - |       - |        - |       - |
| `t_ems_ext_hardhat`               |  2m 05s  |  2m 25s |  2m 56s |  2m 25s |   2m 19s |  2m 30s |
| `t_ems_solcjs`                    |  4m 55s  |  5m 35s |  5m 03s |  4m 57s |   5m 30s |  5m 31s |
| `t_ems_test_ext_ens`              |  4m 43s  |  2m 43s |  2m 45s |  2m 57s |   3m 04s |  2m 41s |
| `t_ems_test_ext_gnosis_v2`        |  3m 16s  |  3m 06s |  3m 31s |  3m 39s |   3m 55s |  3m 40s |
| `t_ems_test_ext_zeppelin`         | 21m 08s  |  7m 59s | 10m 37s |   CRASH |  11m 00s |  9m 29s |
| `t_osx_cli`                       | 16m 12s  | 16m 55s | 16m 03s | 17m 53s |  16m 14s | 16m 25s |
| `t_osx_soltest`                   | 24m 11s  | 23m 09s | 22m 45s | 25m 43s |  27m 45s | 25m 09s |
| `t_ubu_clang_soltest`             | 14m 30s  | 14m 26s | 13m 57s | 15m 32s |  15m 03s | 14m 48s |
| `t_ubu_cli`                       |  7m 58s  |  8m 52s |  8m 34s |  8m 44s |  11m 55s |  9m 43s |
| `t_ubu_pyscripts`                 |     23s  |     27s |     23s |     25s |      22s |     21s |
| `t_ubu_release_cli`               |  8m 12s  |  7m 30s |  9m 05s | 10m 01s |  10m 25s |  8m 37s |
| `t_ubu_release_soltest_all`       | 32m 16s  | 33m 39s | 33m 24s | 33m 10s |  39m 36s | 32m 59s |
| `t_ubu_soltest_all`               | 31m 32s  | 33m 26s | 32m 31s | 32m 06s |  34m 38s | 34m 01s |
| `t_ubu_soltest_enforce_yul`       | 13m 55s  | 15m 05s | 14m 06s | 14m 27s |  14m 58s | 17m 10s |
| `t_win_pyscripts`                 |     46s  |     44s |     48s |  1m 53s |      49s |     43s |
| `t_win_release_soltest`           | 10m 52s  | 10m 47s | 10m 52s | 12m 35s |  11m 59s | 11m 49s |
| `t_win_soltest`                   | 10m 58s  | 10m 54s | 10m 47s | 12m 24s |  11m 50s | 11m 46s |

Suprisingly, the results do not align completely with what I have seen in https://github.com/ethereum/solidity/issues/11670#issuecomment-954297346. In some cases here a `large` machine does increase running time significantly compared to `xlarge`. Maybe in some cases the processes are actually blocked on I/O often and having more of them would help? I originally used `MAKEFLAGS=-j10` on `large` in my tests and maybe this made the difference. I need to investigate that.

EDIT: Added timing from runs I did later:
- (2) run after a rebase.
- (3) run after https://github.com/ethereum/solidity/pull/12215#issuecomment-962039720
- (4) adjusted thread numbers: https://github.com/ethereum/solidity/pull/12215#issuecomment-962178656

**Comment by cameel** [comment]
As for how it affects our credit usage - unfortunately it's hard to say before we merge it. CircleCI does not show the cost per workflow so we'll only see the overall impact on the [Insights page](https://app.circleci.com/insights/github/ethereum/solidity/workflows/main) once this runs for a day or two.

**Comment by cameel** [comment]
Run 3:
- `small` seems to be not enough for `t_ems_test_ext_zeppelin` because it crashed in the second run. Changing to `medium`.
- I'm doubling the number of threads for `b_ubu_clang`, `b_ubu_cxx20`, `b_ubu_static` and `b_ems` without switching to a larger resource class. Let's see if there are some unused CPU cycles after all.

**Comment by cameel** [comment]
Doubling the number of threads indeed helped. For `b_ubu_static` that does not fit in memory though. I did another run with modified numbers.

Run 4:
- `b_ubu_clang`: 2 more threads
- `b_ubu_cxx20`: 2 more threads
- `b_ubu_static`: 1 less thread to prevent it running out of memory
- `t_ems_test_ext_zeppelin`: increased resource class to `large`

The result is that the 2 extra threads do not seem to be doing anything but 1 less for `b_ubu_static` does prevent the crash. `t_ems_test_ext_zeppelin` did complete faster but that might be just normal variance.

Table updated with new numbers.

**Comment by cameel** [comment]
#12214 has been merged so this is now reviewable.

This PR removes three jobs that are marked as required: `t_ems_compile_ext_gnosis_v2`, `t_ems_compile_ext_zeppelin`, `t_ems_compile_ext_ens`. They need to be removed from the repo settings as well before we can merge it.

**Review by chriseth — APPROVED** [review]

**Pull request merged**
